examples/elixir_shallowwater_cubed_sphere_shell_EC_correction.jl

```diff
@@ -0,0 +1,168 @@
+# form obtained through an entropy correction term
+
+equations = ShallowWaterEquations3D(gravity_constant = 9.81)
+
+# Create DG solver with polynomial degree = 3 and Wintemeyer et al.'s flux as surface flux
+polydeg = 3
+volume_flux = flux_fjordholm_etal #flux_wintermeyer_etal
+surface_flux = flux_fjordholm_etal #flux_wintermeyer_etal  #flux_lax_friedrichs
+solver = DGSEM(polydeg = polydeg,
+               surface_flux = surface_flux,
+               volume_integral = VolumeIntegralFluxDifferencing(volume_flux))
+
+# Initial condition for a Gaussian density profile with constant pressure
+# and the velocity of a rotating solid body
+function initial_condition_advection_sphere(x, t, equations::ShallowWaterEquations3D)
+    # Gaussian density
+    rho = 1.0 + exp(-20 * (x[1]^2 + x[3]^2))
+
+    # Spherical coordinates for the point x
+    if sign(x[2]) == 0.0
+        signy = 1.0
+    else
+        signy = sign(x[2])
+    end
+    # Co-latitude
+    colat = acos(x[3] / sqrt(x[1]^2 + x[2]^2 + x[3]^2))
+    # Latitude (auxiliary variable)
+    lat = -colat + 0.5 * pi
+    # Longitude
+    r_xy = sqrt(x[1]^2 + x[2]^2)
+    if r_xy == 0.0
+        phi = pi / 2
+    else
+        phi = signy * acos(x[1] / r_xy)
+    end
+
+    # Compute the velocity of a rotating solid body
+    # (alpha is the angle between the rotation axis and the polar axis of the spherical coordinate system)
+    v0 = 1.0 # Velocity at the "equator"
+    alpha = 0.0 #pi / 4
+    v_long = v0 * (cos(lat) * cos(alpha) + sin(lat) * cos(phi) * sin(alpha))
+    v_lat = -v0 * sin(phi) * sin(alpha)
+
+    # Transform to Cartesian coordinate system
+    v1 = -cos(colat) * cos(phi) * v_lat - sin(phi) * v_long
+    v2 = -cos(colat) * sin(phi) * v_lat + cos(phi) * v_long
+    v3 = sin(colat) * v_lat
+
+    return prim2cons(SVector(rho, v1, v2, v3, 0), equations)
+end
+
+# Source term function to apply Lagrange multiplier to the semi-discretization
+# The vector contravariant_normal_vector contains the normal contravariant vectors scaled with the inverse Jacobian
+function source_terms_lagrange_multiplier(u, du, x, t,
+                                          equations::ShallowWaterEquations3D,
+                                          contravariant_normal_vector)
+
+    # Entropy correction term
+    s2 = -contravariant_normal_vector[1] * equations.gravity * u[1]^2
+    s3 = -contravariant_normal_vector[2] * equations.gravity * u[1]^2
+    s4 = -contravariant_normal_vector[3] * equations.gravity * u[1]^2
+
+    # Lagrange multipliers in the x direction
+    x_dot_div_f = (x[1] * (du[2] + s2) + x[2] * (du[3] + s3) + x[3] * (du[4] + s4)) /
+                  sum(x .^ 2)
+    s2 += -x[1] * x_dot_div_f
+    s3 += -x[2] * x_dot_div_f
+    s4 += -x[3] * x_dot_div_f
+
+    return SVector(0.0, s2, s3, s4, 0.0)
+end
+
+# Custom RHS that applies a source term that depends on du (used to convert apply Lagrange multiplier)
+function rhs_semi_custom!(du_ode, u_ode, semi, t)
```

Comment:
Is it possible to have the custom RHS and Lagrange multiplier source terms defined in `dg_2d_manifold_in_3d_cartesian.jl` and called by dispatch on the `ShallowWaterEquations3D` equation type rather than specified in the elixir? This is how I do it for the covariant form, for example:

https://github.com/trixi-framework/TrixiAtmo.jl/blob/7d7ec99b45eef536b738f9461967b4a6cb350624/src/solvers/dgsem_p4est/dg_2d_manifold_in_3d_covariant.jl#L6-L64

Reply:
This is a good point. I agree with you that it is not optimal to define the custom RHS in every elixir. I implemented it in this way in Trixi (before porting the routines to TrixiAtmo) because the main Trixi developers were reluctant to add "source" terms that depend on and change `du`.

I would prefer to not dispatch on `ShallowWaterEquations3D` inside the solver routines to keep the physics and numerics separated. One possible workaround would be to define a new RHS function that dispatches on `mesh::P4estMesh{2}` and `equations::AbstractEquations{3}`, which uses source terms that depend on and change `du`. Does that sound like a good solution?

Reply:
> One possible workaround would be to define a new RHS function that dispatches on mesh::P4estMesh{2} and equations::AbstractEquations{3}, which uses source terms that depend on and change du. Does that sound like a good solution?

Yes, this approach is what I would do.

Reply:
Regarding your question, I don't know if there is a way to call an unspecialized method from within a specialized method of the same function.  Although there may be a way to work around that, I am not sure duplicating `rhs!` is really a huge issue terms of maintainability. It is a bit ugly, sure, but there is already a lot of duplicated (or repeated with minor differences) code in Trixi.jl for all the solver types and spatial dimensions. I get the impression that this repetition was a design decision in Trixi.jl based on the idea of favouring readability and performance over conciseness, and thus writing a new `rhs!` method for what is effectively a new solver type seems to fit that paradigm.